Recipe for building a toolbar app (#7922)
withastro/docs: src/content/docs/en/recipes/making-toolbar-apps.mdx

---
title: Creating a dev toolbar app
description: Learn how to create a dev toolbar app for your site.
type: recipe
---

import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'
import { FileTree } from '@astrojs/starlight/components';

> Astro includes a [development toolbar](https://docs.astro.build/en/guides/dev-toolbar/) that you can use to inspect your site, check for accessibility and performance issues, and more. 

Reviewer: ```suggestion
Astro includes a [development toolbar](/en/guides/dev-toolbar/) that you can use to inspect your site, check for accessibility and performance issues, and more. 
```

    ```
  </Fragment>
</PackageManagerTabs>

Alternatively, keep on reading to learn how to build an app from scratch.
:::

## Creating the integration

Dev toolbar apps can only be added by [Astro Integrations](/en/guides/integrations-guide/). To create an integration, create a file anywhere in your project exporting an object with the integration name and hooks. 

The `astro:config:setup` hook has access to the `addDevToolbarApp` function, which can be used to add dev toolbar apps. The `addDevToolbarApp` function takes an object with the id, name, icon and finally entrypoint of the dev toolbar app. The entrypoint is the path to the file that exports the dev toolbar app.

```ts title="my-integration.ts"
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import type { AstroIntegration } from 'astro';

export default {
  name: 'my-astro-integration',
  hooks: {
    'astro:config:setup': ({ addDevToolbarApp }) => {
      const __filename = fileURLToPath(import.meta.url);
      const __dirname = dirname(__filename);
  
      addDevToolbarApp({
        id: "my-toolbar-app",
        name: "My Toolbar App",
        icon: "🚀",

>         entrypoint: join(__dirname, "./app.ts"),

Reviewer 2: All the path code doesn't look great. I don't know why we left all of these entrypoints as always being string. Can we at least clear this up a little by reducing to `fileURLToPath(new URL('./app.ts', import.meta.url))` ?

      canvas.appendChild(h1);
    },
});
```

In most cases, the `init()` function will contain the vast majority of your dev toolbar app's logic, as this function will create the UI for your app, set up event listeners, etc. In order to make it easier to do common operations, `init()` also includes two additional arguments: `app` and `server`.

```ts title="app.ts" ins=/(app),/ ins=/(server)\\)/
import { defineToolbarApp } from "astro/toolbar";

export default defineToolbarApp({
    init(canvas, app, server) {
      // Create a new text node every time the app is toggled on or off
      app.onAppToggled(({ state }) => {
        const text = document.createTextNode(
          `The app is now ${state ? "enabled" : "disabled"}!`,
        );
        canvas.appendChild(text);
      });
      
      // Send a message to the server
      server.send("my-app:hello-server", {message: "Hello, server!"});
    },
});
```

These arguments provide multiple utilities and helpers to interact with the dev toolbar and the server, such as executing code when the app is toggled on or off, showing a notification dot on the app, sending messages to the server, and more.

> For more information on the hooks and helpers available in the `init()` function, see the [Astro Toolbar API documentation](https://docs.astro.build/api/astro/toolbar).

Reviewer: ```suggestion
For more information on the hooks and helpers available in the `init()` function, see the [Astro Toolbar API documentation](/en/reference/dev-toolbar-app-reference/).
```
Now you're just making things up!  :smile:

Author: I put it as a placeholder while writing and forgot to change it, ha

    },
});
```

The [HTM package](https://github.com/developit/htm) is a good choice for creating dev toolbar apps without a build step, offering native integration for React and Preact and support for other frameworks.

```ts title="app.ts" ins={3, 7}
import { defineToolbarApp } from "astro/toolbar";
import { render } from "preact";
import { html } from 'htm/preact';

export default defineToolbarApp({
    init(canvas) {
        render(html`<h1>Hello World!</h1>`, canvas);
    },
});
```

### With a build step

Astro does not preprocess JSX code in dev toolbar apps. As such, a build step is required in order to use JSX components in your dev toolbar app. In this recipe, we'll be using TypeScript to do this, but any other tools that compile JSX code will work as well (e.g. Babel, Rollup, ESBuild).

First, install TypeScript inside your project:

<PackageManagerTabs>
  <Fragment slot="npm">
    ```shell
    npm install --save-dev typescript
    ```
  </Fragment>
  <Fragment slot="pnpm">
    ```shell
    pnpm install --save-dev typescript
    ```
  </Fragment>
  <Fragment slot="yarn">
    ```shell
    yarn add --dev typescript
    ```
  </Fragment>
</PackageManagerTabs>

> Next, create a `tsconfig.json` file in the root of your project with the appropriate settings to build and for the framework you're using ([React](https://react-typescript-cheatsheet.netlify.app/docs/basic/setup), [Preact](https://preactjs.com/guide/v10/typescript), [Solid](https://www.solidjs.com/guides/typescript)). For example, for Preact:

Reviewer 2: I read "root of your project" as top-level. Would "root of your toolbar app" make it more clear?

Create a new file called `app.ts` in the same directory as your integration file and export the result of calling `defineToolbarApp` with an object containing an `init` function. We'll leave the `init` function empty for now:

```ts title="app.ts"
import { defineToolbarApp } from "astro/toolbar";

export default defineToolbarApp({
    init() {
      // ...
    },
});
```

The `init()` function receives a standard [ShadowRoot](https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot) as its first argument, which you can use to create a UI for your dev toolbar app. Elements can be created and added to the ShadowRoot using the standard DOM APIs.

Create a new `h1` element with a motivational message and append it to the ShadowRoot:

```ts title="app.ts" ins={3-8, 12-15}
import { defineToolbarApp } from "astro/toolbar";

const motivationalMessages = [
  "You're doing great!",
  "Keep up the good work!",
  "You're awesome!",
  "You're a star!",
];

export default defineToolbarApp({
    init(canvas) {
      const h1 = document.createElement('h1');
      h1.textContent = motivationalMessages[Math.floor(Math.random() * motivationalMessages.length)];

>       canvas.appendChild(h1);

Reviewer 2: ```suggestion
      canvas.append(h1);
```

    export default {
      name: 'my-astro-integration',
      hooks: {
        'astro:config:setup': ({ addDevToolbarApp }) => {  
          addDevToolbarApp({
            id: "my-toolbar-app",
            name: "My Toolbar App",
            icon: "🚀",
            entrypoint: fileURLToPath(new URL('./app.ts', import.meta.url))
          });
        },
      },
    } satisfies AstroIntegration;
    ```

    :::note[Using relative paths to the entrypoint]
    The `entrypoint` is the path to your dev toolbar app file **relative to the root of your existing Astro project**, not to the integration folder (`my-toolbar-app`) itself.

    To use relative paths for entrypoints, get the path to the current file using `import.meta.url` and resolve the path to the entrypoint from there.
    :::

3. To use this integration in your project, add it to the `integrations` array in your `astro.config.mjs` file.

    ```js title="astro.config.mjs" ins={2,5}
    import { defineConfig } from 'astro/config';
    import myIntegration from './my-toolbar-app/my-integration.ts';

    export default defineConfig({
      integrations: [myIntegration],
    })
    ```

> 4. If not already running, start the dev server. If your integration has been successfully added to your project, you should see a new "undefined" app in the dev toolbar.

Author: Hmm, does it really show as undefined? That's not intended

Reviewer: It did for me, yes! Just a question mark for the icon, and title "undefined" when I hovered over it. I built a couple of different times from scratch and saw this behaviour both times. If it's not intended, and it will work to show the icon (just nothing happens when you click) then of course we can go back to the original wording!

I wondered whether it made sense to build the app first, before the integration to also avoid the console error, but this also matches the order of things on the API page, and I like how it emphasizes that you're building an integration.

Author: Aah I see yes, it does that when the file doesn't exist at all. An empty file would probably work so it at least shows the icon and title correctly.

Reviewer: I tried with an empty file too, and got the same result!
